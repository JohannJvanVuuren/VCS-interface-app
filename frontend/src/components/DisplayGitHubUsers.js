/* Import of the main stylesheet generated by the SCSS files and preprocessor */
import '../sass/main.css';

/* Import of the useLocation hook which will enable to get the state passed via the Link components in
* the Header component */
import {useLocation} from 'react-router-dom';
import {useEffect, useState} from "react";
import ReactPaginate from 'react-paginate';


/* Definition of the DisplayGitHubUsers component */
export const DisplayGitHubUsers = () => {

    /* Declaration and initialisation of state variables */
    const [users, setUsers] = useState([]);
    const [isLoading, setIsLoading] = useState(false);
    const [currentUsers, setCurrentUsers] = useState(null);
    const [pageCount, setPageCount] = useState(0);
    const [usersOffset, setUsersOffset] = useState(0);

    /* Setting the user that will be displayed per page. This can be changed at a later stage so that the
    * value is obtained from the user */
    const usersPerPage = 10;

    /* Creation of an instance of useLocation */
    const location = useLocation();
    /* Storing the state data from the useLocation instance */
    const data = location.state.gitHubUserArray;



    useEffect(() => {
        const endOffset = usersOffset + usersPerPage;
        setCurrentUsers(data.slice(usersOffset, endOffset));
        setPageCount((Math.ceil(data.length / usersPerPage)))
    }, [data, usersOffset]);

    const pageClickHandler = (event) => {
        const newOffset = (event.selected * 8) % data.length;
        setUsersOffset(newOffset);
    }

    return (
        <div className={'github-table'}>
            <h1>GitHub User Listing</h1>
            <table className={'gitHubTable'}>
                <thead>
                <tr>
                    <th>Username</th>
                    <th>Repo URL</th>
                    <th>Avatar</th>
                </tr>
                </thead>
                <tbody>
                {data.map((item, index) => {
                    return (
                        <tr key={index}>
                            <td>{item.login}</td>
                            <td>{item.repos_url}</td>
                            <td>{item.avatar_url}</td>
                        </tr>
                    )
                })}
                </tbody>
            </table>
        </div>
    )

}