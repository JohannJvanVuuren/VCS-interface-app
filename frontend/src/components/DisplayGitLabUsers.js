/* Import of the main stylesheet generated by the SCSS files and the preprocessor */
import '../sass/main.css';

/* Import of the useLocation hook which will enable to get the state passed via the Link components in
* the Header component */
import { useLocation } from 'react-router-dom';

/* Definition of the DisplayGitLabUsers component */
export const DisplayGitLabUsers = () => {

    /* Creation of an instance of useLocation */
    const location = useLocation();
    /* Storing the state data from the useLocation instance */
    const data = location.state.gitLabUserArray;

    console.log(data);

    return(
        <div>
            <h2 className="mb-4">
                React Send / Get State using Link Component Example
            </h2>
            <p>Login: {data[0].username}</p>
        </div>
    )
}