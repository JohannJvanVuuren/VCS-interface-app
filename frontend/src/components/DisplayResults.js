import '../sass/main.css';

export const DisplayResults = () => {

}