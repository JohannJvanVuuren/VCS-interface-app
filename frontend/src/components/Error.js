/* Import of main stylesheet generated from SCSS and preprocessor */
import '../sass/main.css';

/* Component to display an error message when the user enters an unknown route */
export const Error = () => {
    return (
        <main className={'error'}>
            <h1>Oops! Page not found!</h1>
        </main>
    )
};
