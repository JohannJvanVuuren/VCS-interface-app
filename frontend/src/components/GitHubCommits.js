/* Import of main stylesheet generated by the SCSS files and preprocessor*/
import '../sass/main.css';
/* Import of axios client to make API call to backend */
import axios from "axios";
/* Import of React Hooks used in this component */
import { useEffect, useState } from "react";
import { useLocation } from 'react-router-dom';

/* Definition of the GitHubCommits component that displays all commits for a specific repo. The commits are in
* descending order so the latest commit will appear at the top of the list */
export const GitHubCommits = () => {

    /* Declaration and initialisation of the state variable needed in this component */
    const [commitsList, setCommitsList] = useState([]);

    /* Use of the useLocation hook to get the state attribute data from the Link component to this page */
    const location = useLocation();
    const user = location.state.user;
    const repoName = location.state.repoName;

    /* The useEffect hook to make the axios call to the backend only when the user or repoName changes */
    useEffect(() => {

        /* The axios call to the githubCommits route on the backend. The username and repo name is encoded in
        * the body of the post request */
        axios.post('http://localhost:8000/githubCommits', {
            user: user,
            repoName: repoName
        })
            .then(response => {
                setCommitsList(response.data)
            })
            .catch(error => {
                console.log('Error', error.message);
            })

    }, [user, repoName]); // useEffect hook's dependencies

    return (
        <div className={'commit-container'}>
            {/* Using the Array.map method to loop through the list of commits received back from the backend
             and render the required properties */}
            {commitsList.map((commit, index) => {
                return (
                    <div key={index}>
                        <p>Date: {commit.commit.author.date} <span>Comment: {commit.commit.message}</span></p>
                    </div>
                )
            })}
        </div>
    )
}