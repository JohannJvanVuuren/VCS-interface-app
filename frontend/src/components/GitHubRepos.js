/* Import of the main stylesheet generated by the SCSS files and preprocessor */
import '../sass/main.css';
import Spinner from 'react-bootstrap/Spinner';

/* Import of useLocation hook to get the state information from the Link component to this component */
import {useLocation, Link} from 'react-router-dom';
import {useEffect, useState} from "react";
import axios from 'axios';

export const GitHubRepos = () => {

    const [repos, setRepos] = useState([]);
    const [isLoading, setIsLoading] = useState(false);

    const location = useLocation();
    const user = location.state.user.login;

    useEffect(() => {

        setIsLoading(true);

        axios.post('http://localhost:8000/githubRepos', {
            user: user
        })
            .then(response => {
                setRepos(response.data);
                setIsLoading(false);
                return response.data;
            })
            .catch(error => {
                console.log(error);
                setIsLoading(false);
            })


    }, [user]);

    return (
        <div>
            <h1>{user}</h1>
            <h2>Repositories</h2>
            {isLoading ?
                <div className={'mx-auto'}>
                    <Spinner animation="border" variant='light'/>
                </div>

             :
                repos.length === 0 ?

                    <h2>No records found</h2>

                :
                repos.map((repo, index) => {
                    return (
                        <div className={'repo-wrapper'} key={index}>
                            <p>{repo.name}</p>
                            <Link
                                to={'/gitHubCommits'}
                                state={{
                                    repoName: repo.name,
                                    user: user
                                }}
                            >
                                Commits
                            </Link>
                        </div>
                    )
                })}

        </div>
    )
}