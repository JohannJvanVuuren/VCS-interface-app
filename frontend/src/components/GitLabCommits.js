import '../sass/main.css';
import { useEffect, useState } from "react";
import { useLocation } from 'react-router-dom';
import axios from "axios";

export const GitHubCommits = () => {

    const [commitsList, setCommitsList] = useState([]);

    const location = useLocation();
    const repoId = location.state.repoId;

    useEffect(() => {

        axios.post('http://localhost:8000/githubCommits', {
            repoId: repoId
        })
            .then(response => {
                setCommitsList(response.data)
            })
            .catch(error => {
                console.log(error);
            })

    }, [repoId]);

    return (
        <div>
            {commitsList.map((commit, index) => {
                return (
                    <div>
                        <p>Date: {commit.comitted_date} <span>Comment: {commit.title}</span></p>
                    </div>
                )
            })}
        </div>
    )
}