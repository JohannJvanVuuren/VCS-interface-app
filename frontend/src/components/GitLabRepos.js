/* Import of the main stylesheet generated by the SCSS files and preprocessor */
import '../sass/main.css';
/* Import of the Spinner component from React-Bootstrap */
import Spinner from 'react-bootstrap/Spinner';

/* Import of the axios client to handle the call to the backend API */
import axios from 'axios';

/* Import of the React and React Router hooks needed in this component */
import {useLocation, Link} from 'react-router-dom';
import {useEffect, useState} from "react";

/* The definition of the GitLabRepos component. This component makes a post request to the gitlabRepos
* route on the backend and uses the data received back to display the repos for a specific user. */
export const GitLabRepos = () => {

    /* The declaration and initialisation of the state variables needed in this component */
    const [repos, setRepos] = useState([]);
    const [isLoading, setIsLoading] = useState(false);

    /* Making use of the useLocation hook to get the user, webUrl and id information from the Link component that
    * linked out to this page from DisplayGitLabUsers */
    const location = useLocation();
    const user = location.state.user.username;
    const id = location.state.user.id;
    const webUrl = location.state.webUrl;

    /* The useEffect hook to handle the axios request to the backend whenever the user and dependency changes */
    useEffect(() => {

        /* This variable setting is used to make sure a spinner is displayed while the data is being fetched
        *  from the backend */
        setIsLoading(true);

        /* The axios post request to the backend route. The username is encoded in the body of the request  */
        axios.post('http://localhost:8000/gitlabRepos', {
            user: user,
            id: id,
            repoId: id
        })
            .then(response => {
                const repoList = response.data;
                setRepos(repoList);
                setIsLoading(false);
                return response.data;
            })
            .catch(error => {
                console.log('Error', error.message);
                setIsLoading(false);
            })


    }, [user, id]); // useEffect hook dependency


    return (
        <div className={'repo-container'}>
            <div className={'heading'}>
                <div>
                    {/* This links out to the user's GitLab web URL when the username is clicked */}
                    <a href={webUrl} target={'_blank'} rel="noreferrer">
                        <h1 className={'heading-primary'}>{user}</h1>
                    </a>
                </div>
                <h2>Repositories</h2>
            </div>
            <div>
                {/* This ternary operator handles the logic of displaying a spinner while the data is being
                 fetched from the backend. It is a chained ternary operator to handle the case where no repos
                 are found */}
                {isLoading ?
                    <div className={'mx-auto'}>
                        <Spinner animation="border" variant='light'/>
                    </div>
                    :
                    repos.length === 0 ?
                        <h2 className={'no-records'}>No records found</h2>
                        :
                        /* The display of the repo names together with a link to the commits associated with
                        the specific repo */
                        <div className={'repo-flexbox'}>
                            {repos.map((repo, index) => {
                                return (
                                    <div className={'repo-wrapper'} key={index}>
                                        <a href={`https://gitlab.com/${user}/files/${repo.name}`} target={'_blank'}
                                           rel="noreferrer">
                                            <p>{repo.name}</p>
                                        </a>
                                        {/* The repo name, repo ID and usernames are being passed onto the
                                          * gitHubCommits route via the state attribute of the React Router
                                          * Link component here */}
                                        <Link
                                            to={'/gitHubCommits'}
                                            state={{
                                                repoName: repo.name,
                                                repoId: repo.id,
                                                user: user
                                            }}
                                        >
                                            Commits
                                        </Link>
                                    </div>
                                )
                            })}
                        </div>
                }
            </div>
        </div>
    )
}