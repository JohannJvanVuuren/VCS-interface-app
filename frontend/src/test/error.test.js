import renderer from 'react-test-renderer';

import { Error } from '../components/Error';

it('Error component renders correctly', () => {
    const tree = renderer.create(<Error/>).toJSON();
    expect(tree).toMatchSnapshot();
})