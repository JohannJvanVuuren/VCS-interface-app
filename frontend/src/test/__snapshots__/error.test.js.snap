// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error component renders correctly 1`] = `
<main
  className="error"
>
  <h1>
    Oops! Page not found!
  </h1>
</main>
`;
